CMake: Enable improved armclang support in CMake 3.21

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

When using `armclang`, we normally need to specify the CPU core (e.g. cortex-m33) and any features changes (e.g. no DSP, no floating points, etc.) in CMake _compile_ options only. The linker `armlink` can automatically figure out all the information from object files, so we don't need to specify them again.

There's a problem in CMake versions 3.20 and below: they always add both compiler and linker flags based on `CMAKE_SYSTEM_PROCESSOR`, for example `--cpu=Cortex-M33`, and projects have no way to tell CMake to skip them. This sometimes overrides features in objects files, and results in link failure or unbootable binaries. For example, object files are compiled with no floating points, but the linker's `--cpu=Cortex-M33` assumes floating points. To workaround this, we added more linker flags to "counter" what CMake automatically adds. Example (note: the comment is not entirely accurate): https://github.com/ARMmbed/mbed-os/blob/d147abc3e556c58e5e343d34b729bc2192e18bd3/tools/cmake/cores/Cortex-M33.cmake#L15-L21.

From CMake 3.21 onwards, CMake by default (policy CMP0123 = NEW) does not add flags to `armclang` or `armlink` anymore, and it fully relies on projects to set all flags. In this case we do not need to set `CMAKE_SYSTEM_PROCESSOR` or use workarounds anymore. We still conditionally turn on workarounds when an older version of CMake (3.19 and 3.20) is used, but in the future we might require users to have at least CMake 3.21.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

Users are advised to upgrade CMake to 3.21 (now released), but for now we keep supporting CMake 3.19 and 3.20.

### Documentation <!-- Required -->

None

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Locally tested with this PR + the change in the CMake project.
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core 

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
CMake: Enable improved armclang support in CMake 3.21

CMake versions 3.20 and below always add compiler and linker flags
based on `CMAKE_SYSTEM_PROCESSOR`, for example `--cpu=Cortex-M33`.
This sometimes overrides flags we set in tools/cmake/cores/*.cmake
and results in link failure or unbootable binaries. To workaround
this, we added more linker flags to "counter" what CMake automatically
adds.

From CMake 3.21 onwards, CMake by default does not add flags to
armclang anymore, and it fully relies on projects to set all flags.
In this case we do not need to set `CMAKE_SYSTEM_PROCESSOR` or use
workarounds anymore. We still turn on workarounds when an older
version of CMake (3.19 and 3.20) is used, but in the future we might
require users to have at least CMake 3.21.

hugueskamba commented on tools/cmake/app.cmake:
```diff
@@ -1,6 +1,9 @@
 # Copyright (c) 2020 ARM Limited. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
+# Enable fix for the Arm Compiler: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6035
```
Can the comment here instead (or also) describes that this policy is about?

LDong-Arm replied:
This is temporary and serves to force enable an unreleased feature. It'll be removed once CMake is updated. I'll make this PR ready for review then.

LDong-Arm replied:
It should be part of 3.21.0, hopefully it'll be out soon - then we can bump `cmake_minimum_required`.

wernerlewis commented on tools/cmake/cores/Cortex-M4.cmake:
```diff
@@ -12,11 +12,13 @@ elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
+        list(APPEND link_options
+            "--cpu=Cortex-M4.no_fp"
+        )
+    if(deprecated_system_processor)
```
Minor correction for this if block.
```suggestion
    endif()
```

wernerlewis commented on tools/cmake/mbed_toolchain.cmake:
```diff
@@ -22,47 +22,75 @@ function(mbed_generate_options_for_linker target output_response_file_path)
+# Note: From version 3.21, CMake by default (policy CMP0123 set to NEW) does not
+# use this macro anymore, and projects have full control over compile and link
+# options. This is because the old algorithm based on CMAKE_SYSTEM_PROCESSOR
+# is too restrictive and doesn's support things like Cortex-M33.no_dsp.no_fp.
```
```suggestion
# is too restrictive and doesn't support things like Cortex-M33.no_dsp.no_fp.
```

LDong-Arm committed:
CMake: Cortex-A: Use MBED_CPU_CORE to set -mcpu

As we're moving away from the use of CMAKE_SYSTEM_PROCESSOR, use
MBED_CPU_CORE instead. They differ in cases, but both GCC and Arm
toolchains support case-insensitive CPU names.

LDong-Arm committed:
CMake: Do not set `CMAKE_AR`

The variable for archiver is automatically set by CMake. There is
no need to explicitly set it, unless a project really needs to
override the default archiver.
